src/librustc/ty/mod.rs

```diff
@@ -2623,3 +2575,47 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         }
     }
 }
+
+fn associated_item<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
```

Comment:
I opened up https://github.com/rust-lang/rust/issues/40697 where we can discuss the precise plan. It feels a bit like this should go into a follow-up PR to me; this PR is just converting to on-demand but keeping the existing structure.